packages/datepicker-react/MIGRATION.md

```diff
@@ -0,0 +1,111 @@
+
+Om resten av koden din fungerer best med `Date` så er dette den minimale endringen du trenger å gjøre for å få satt riktig `value`:
+
+```diff
++ import { DatePicker, formatInput } from "@fremtind/jkl-datepicker-react";
+
+  <DatePicker
+-    value={value}
++    value={formatInput(value)}
+  />
+```
+
+### Endring i output
+
+Endringen i rekkefølge på eventlyttere gjør at `register()` fungerer med React Hook Form.
+
+Om resten av koden din fungerer best med `Date` så er dette den minimale endringen du trenger å gjøre i `onChange`:
+
+```diff
+  <DatePicker
+    value={formatInput(value)}
+-   onChange={(date, event, meta) => {
++   onChange={(event, date, meta) => {
+      setValue(date);
+    }}
+  />
+```
+
+### Endringer i `onBlur`
+
+Hvis du bruker `onBlur` til å gjøre noe i dag er det endringer som treffer deg.
+
+tl;dr – du må ha `onBlur` to steder – på inputfeltet og på kalenderknappen (`action`) – og bruke `isBlurTargetOutside`:
+
+```tsx
+import { DatePicker, isBlurTargetOutside } from "@fremtind/jkl-datepicker-react";
+
+<DatePicker
+    label="Velg startdato for forsikringen"
+    value={value}
+    onBlur={(e, date, meta) => {
+        if (isBlurTargetOutside(e)) {
+            // Brukeren har forlatt DatePicker. Gjør noe her.
+        } else {
+            // Ignorer blurs som går til kalenderknapper
+        }
+    }}
+    action={{
+        onBlur: (e) => {
+            if (isBlurTargetOutside(e)) {
+                // Brukeren har forlatt DatePicker. Gjør noe her.
+            } else {
+                // Ignorer blurs som går tilbake til inputfeltet
+            }
+        },
+    }}
+/>;
+```
+
+#### OK, men hvorfor?
+
+`blur` er et global event som skjer hver gang et fokuserbart element _mister_ fokus. `DatePicker` består av flere fokuserbare elementer.
+
+De to viktigste fokuserbare elementene i `DatePicker` er _inputfeltet_ og _kalenderknappen_. Hvis en tastaturbruker trykker `Tab` gjennom en `DatePicker` vil det alltid skje to `blur`-events i nettleseren:
+
+-   `onBlur` på inputfeltet, fokus flytter seg til kalenderknappen
+-   `onBlur` på kalenderknappen, fokus har forlatt `DatePicker`
+
+Men, en bruker kan åpne kalenderen. Da vil `onBlur` skje på kalenderknappen, men fokus er fortsatt i `DatePicker` – fokuset er flyttet til kalenderen. Derfor har vi laget en utility-funksjon `isBlurTargetOutside` som sjekker i `blur`-eventet hvor fokus har havnet. Om vi ser at det er på et element i `DatePicker` så svarer den `false`.
+
+`DatePicker` er kodet sånn at fokus alltid flyttes tilbake til enten inputfeltet eller kalenderknappen etter at kalenderen har vært aktiv. På den måten får vi alltid minst ett `blur`-event hvis brukeren har vært innom `DatePicker`.
+
+Grunnen til at vi ikke ønsker å samle disse to handlerene under samme `onBlur` er at det er uventet for de fleste å motta to `onBlur`-events i den samme handleren, fra ulike kilder.
```

Comment:
Kunne dette vært forenklet ved å stoppe propagering av enkelte blurs?

Reply:
Da blir det et spørsmål om timing. Hvilke av blur-eventene skal vi sluke? Skal vi for eksempel aldri sende blurs for kalenderknappen? I så fal, når skal vi sende blur-event for inputfeltet? Tar vi vare på det i state og sender det senere?

Jeg tror det er en _can of worms_ vi helst ikke vil deale med i fremtiden.

Jeg vet dette APIet er litt herk, men jeg tror dette er _the lesser evil_ fra et rammeverkperspektiv. Kan hende det lønner seg å lage en wrapperkomponent i egen app @BjorneOmaFremtind som skjuler litt av kompleksiteten. Vet dere gjør ting på blur i SeOpp.